### PR TITLE
Add EIP-712 encoding for multiple messages of the same type

### PR DESCRIPTION
**Pull request by austinchandra**
Closes: #[ENG-929](https://linear.app/evmos/issue/ENG-929/add-support-for-multiple-messages-in-one-tx-in-eip-712-verification)

## Description

- Adds support for EIP-712 encoding a transaction with multiple messages, provided they are all of the same type
- Includes minor refactors + renames for better clarity

This PR changes code in `eip712/encoding.go`, which is used within `ethsecp256k1.go` when verifying signatures. The new code enforces different restrictions on the payload messages—there can be multiple messages, but they must all have the same `type`. Due to its typing design, EIP-712 will always be incompatible with payloads with multiple messages of different types. Thus, we ensure that only single-typed instances can pass verification. 

For the time being, only single-signer payloads are supported. This will be changed in the future.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/eip712/encoding.go [hunk 1]
```diff
@@ -219,3 +221,67 @@ func decodeProtobufSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
...
+func validatePayloadMessages(msgs []sdk.Msg) error {
+	if len(msgs) == 0 {
+		return errors.New("unable to build EIP-712 payload: transaction does contain any messages")
+	}
+
+	var msgType string
+	var msgSigner sdk.AccAddress
+
+	for i, m := range msgs {
+		t, err := getMsgType(m)
```
Comment (fedekunze) [inline ethereum/eip712/encoding.go:236]: rename variable `t` for readibility

File: ethereum/eip712/encoding.go [hunk 2]
```diff
@@ -219,3 +221,67 @@ func decodeProtobufSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
...
+			return errors.New("unable to build EIP-712 payload: expect exactly 1 signer")
+		}
+
+		if i == 0 {
+			msgType = t
+			msgSigner = m.GetSigners()[0]
+			continue
+		}
+
+		if strings.Compare(t, msgType) != 0 {
```
Comment (fedekunze) [inline ethereum/eip712/encoding.go:251]: nit, an `==` would be more suitable here

File: ethereum/eip712/encoding.go [hunk 3]
```diff
@@ -219,3 +221,67 @@ func decodeProtobufSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
...
+// getMsgType returns the message type prefix for the given Cosmos SDK Msg
+func getMsgType(msg sdk.Msg) (string, error) {
+	jsonBytes, err := aminoCodec.MarshalJSON(msg)
+	if err != nil {
+		return "", err
+	}
+
+	var jsonMsg aminoMessage
+	err = json.Unmarshal(jsonBytes, &jsonMsg)
+	if err != nil {
```
Comment (fedekunze) [inline ethereum/eip712/encoding.go:277]:
```suggestion
	if err := json.Unmarshal(jsonBytes, &jsonMsg); err != nil {
```

File: ethereum/eip712/encoding.go [hunk 4]
```diff
@@ -219,3 +221,67 @@ func decodeProtobufSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
...
+		}
+	}
+
+	return nil
+}
+
+type aminoMessage struct {
+	Type  string      `json:"type"`
+	Value interface{} `json:"value"`
+}
```
Comment (fedekunze) [inline ethereum/eip712/encoding.go:266]: nit, move types to the top of the file

**Review by fedekunze — APPROVED** [review]
LGTM, minor comments

**Review by Vvaradinov — APPROVED** [review]
LGTM! 
NIT: consider renaming `types2` and `types3` import in `utils_test.go` for better readability.

File: ethereum/eip712/encoding.go (hunk 1 above)
Comment (austinchandra) [reply ethereum/eip712/encoding.go:236]: Any suggestions? Used `t` because it's easy to distinguish from `msgType`

**Pull request merged**